diff -Naur openssl-1.0.2r/apps/Makefile openssl-1.0.2r/apps/Makefile
--- openssl-1.0.2r/apps/Makefile	2019-02-26 15:20:20.000000000 +0100
+++ openssl-1.0.2r/apps/Makefile	2019-04-04 20:36:54.983592324 +0200
@@ -159,10 +159,9 @@
 	elif [ -n "$(FIPSCANLIB)" ]; then \
 	  FIPSLD_CC="$(CC)"; CC=$(FIPSDIR)/bin/fipsld; export CC FIPSLD_CC; \
 	fi; \
-	LIBRARIES="$(LIBSSL) $(LIBKRB5) $(LIBCRYPTO)" ; \
 	$(MAKE) -f $(TOP)/Makefile.shared -e \
 		APPNAME=$(EXE) OBJECTS="$(PROGRAM).o $(E_OBJ)" \
-		LIBDEPS="$(PEX_LIBS) $$LIBRARIES $(EX_LIBS)" \
+		LIBDEPS="$(PEX_LIBS) -L.. -lssl -L.. -lcrypto $(LIBKRB5) $(EX_LIBS)" \
 		link_app.$${shlib_target}
 	@(cd ..; $(MAKE) rehash)
 
